keps/sig-apps/3335-statefulset-slice/README.md

```diff
@@ -683,13 +696,29 @@ rollout. Similarly, consider large clusters and how enablement/disablement
 will rollout across nodes.
 -->
 
+If a control plane rollout disables this feature, the StatefulSet controller
+will update ordinal numbers it controls. This will result in pods being deleted,
+while other pods are scaled in. The StatefulSet controller scales up pods before
+it deletes pods, so as a result, the StatefulSet should not manage fewer than the
+number of replicas that are defined in the spec. Disabling the feature may have an effect on the
+stateful workload that is being run. If the stateful application expects a
+specific ordinal number to be available, it may result in an application failing
+to reach quorum, or rebalancing data based on the number of available replicas.
+
 ###### What specific metrics should inform a rollback?
 
 <!--
 What signals should users be paying attention to when the feature is young
 that might indicate a serious problem?
 -->
 
+The `kube_statefulset_status_replicas` metric can be monitored against the
+`kube_statefulset_replicas` metric to check the expected number of replicas to
+the actual number of pods matched by this StatefulSet's selector. If there is
+a divergence between these fields during steady state operations, this can
+indicate that the number of replicas being created by the StatefulSet do not
+match the expected number of replicas.
```

Comment:
It's probably related with the other question, what kind of metric could be added to inform an admin that there's a mismatch between user settings and what's actually running when ordinal comes to play? I don't think that compering these two metrics give you that, does it?

Reply:
> I don't think that compering these two metrics give you that, does it?
That's correct, in a steady state situation, comparing these two metrics will look the same regardless of the start ordinal.

One option is to emit a new status field (StartOrdinal) and also emit that as a metric. The status `StartOrdinal` and spec `StartOrdinal` could be compared as input/output to verify what the StatefulSet controller is doing. The `.spec.ordinals.start` would be what the StatefulSet should reconcile to, and the `.status.startOrdinal` would be the observed lowest index ordinal pod that is observed with the StatefulSet's selector.

This probably wouldn't be able to be used as an SLO, since reconcilation depends on parameters of a user's workload. Consider the scenario where the `.spec.ordinals.start` is increased from `m` to `n`, where `n` > `m`. There could be scheduling, or grace period reasons why a pod is not created/terminated, leading to differences between `.spec.ordinals.start` and `.status.startOrdinal` for a non-deterministic period of time.

The other benefit of having this `status` field would be an additional assurance to an operator that the StatefulSet is fully reconciled to the correct start ordinal. Currently an operator can check `status.readyReplicas` and `status.replicas` to determine if the right number of replicas are running and ready, however there is no way to check on the status that the ordinal numbers are correct. The operator must rely on the fact that the StatefulSet is correctly reporting `replicas` as the set of replicas that match its selector.

@soltysh What are your thoughts about adding this additional `.status` field, for observability? I think it could be used for correctness for a single StatefulSet, but may be harder to use a cluster level indicator that all the StatefulSets are behaving as expected.

Reply:
I don't think that adding this kind of field at the status level has that much value. I'd be inclined to try to look up something at the cluster-admin level that could help with operational aspects. I'd treat it as nice to have, potentially something we'll need to close before we GA.

Reply:
I believe the `status.observedGeneration` field is supposed to reflect what spec generation the controller has acted on.